integrations/access/accessmonitoring/access_monitoring_rules.go

```diff
@@ -0,0 +1,196 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package accessmonitoring
+
+import (
+	"context"
+	"maps"
+	"slices"
+	"sync"
+
+	"github.com/gravitational/trace"
+
+	accessmonitoringrulesv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/accessmonitoringrules/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/integrations/access/common"
+	"github.com/gravitational/teleport/integrations/access/common/teleport"
+	"github.com/gravitational/teleport/integrations/lib/logger"
+)
+
+const (
+	// defaultAccessMonitoringRulePageSize is the default number of rules to retrieve per request
+	defaultAccessMonitoringRulePageSize = 10
+)
+
+// RuleHandler stores a cache of Access Monitoring Rules for use with Access Request routing in plugins.
+// Must be initialised by calling InitAccessMonitoringRulesCache, a watcher on Acccess Monitoring Rules must pass in new rules using HandleAccessMonitoringRule.
+type RuleHandler struct {
+	accessMonitoringRules AMRMap
+
+	apiClient  teleport.Client
+	pluginType string
+
+	fetchRecipientCallback func(ctx context.Context, recipient string) (*common.Recipient, error)
+}
+
+// AMRMap is a concurrent map for access monitoring rules.
+type AMRMap struct {
+	sync.RWMutex
+	// rules are the access monitoring rules being stored.
+	rules map[string]*accessmonitoringrulesv1.AccessMonitoringRule
+}
+
+// AccessMonitoringRuleHandlerConfig stores the configuration for AccessMonitoringRuleHandler
+type AccessMonitoringRuleHandlerConfig struct {
```

Comment:
```suggestion
type RuleHandlerConfig struct {
```

Similarly, we don't need to say "access monitoring" twice. `accessmonitoring.RuleHandlerConfig` is perfect.